feat(Reactions): added new Reactions component

Ruminat opened the pull request:
### Reactions component

A component for managing users' reactions:

<img width="636" alt="изображение" src="https://github.com/gravity-ui/components/assets/18216279/147c9f88-ab4a-4380-812c-3e360c9e08e8">

<img width="524" alt="изображение" src="https://github.com/gravity-ui/components/assets/18216279/9c4cbf0b-73ec-4be6-903b-4bc7c9ec30af">

ogonkov commented on src/components/Reactions/Reaction.tsx:
```diff
@@ -0,0 +1,112 @@
+    const {value, disabled, selected, content, counter, tooltip} = props.reaction;
+    const {size, onClick} = props;
+
+    const onClickCallback = useStableCallback(() => onClick?.(value));
```
Why it's not just `useCallback`?

ogonkov replied:
I dont get why in this particular case you dont want to use useCallback

Ruminat replied:
`useStableCallback` seems to be a better `useCallback`
But I don't insist on using it here
I'll return the useCallback if you want to

Ruminat replied:
Removed `useStableCallback`

ogonkov commented on src/components/Reactions/Reactions.scss:
```diff
@@ -0,0 +1,49 @@
+        font-size: 10px;
+    }
+    &__reaction-button-text_size_s {
+        font-size: 12px;
```
I guess font-size should be taken from uikit tokens

Ruminat replied:
You mean css-variables like `--g-text-display-1-font-size`?

ogonkov replied:
Yes

Ruminat replied:
Used these font-sizes:
`xs`: `--g-text-caption-1-font-size`
`s`: `--g-text-caption-2-font-size`
`m`: `--g-text-body-1-font-size`
`l`: `--g-text-subheader-1-font-size`
`xl`: `--g-text-subheader-2-font-size`

Ruminat commented:
@ogonkov @korvin89 @amje @ValeraS ping

amje commented:
@Ruminat Let's simplify the API, currently it looks a bit hard to understand and it stores the state in multiple places (`reactions` and `palette.options`). What do you think about this API suggestion?
```ts
// Do not extend whole PaletteOption, only pick needed props. This makes refactor easier
interface ReactionsItem extends Pick<PaletteOption, 'value' | 'content' | 'title'> {
    count: number;
    selected?: boolean;
}

interface ReactionsProps {
    // This would map into Palette props (options and value depending on selected states) and create buttons if count > 0
    reactions: ReactionsItem[];
    // User should handle this callback and change reactions array
    onToggle?: (reaction: ReactionsItem) => void;
    // Extract tooltip render to dedicated prop and pass whole reaction item
    // No other tooltip props needed for the sake of simplicty
    renderTooltipContent?: (reaction: ReactionsItem) => React.ReactNode;
    // Keep only columns. Other Palette props are useless
    columns?: number;
    // Rest props
    // ...
}
```

The only source of state is now the `reactions` array, which user have to update on `onToggle` calls. I also extracted tooltip prop of reaction into top-level prop `renderTooltipContent`. Copying the same tooltip props for each reaction is redundant.

amje commented on src/components/Reactions/Reactions.tsx:
```diff
@@ -0,0 +1,135 @@
+    'value' | 'defaultValue' | 'onUpdate' | 'size' | 'disabled' | 'multiple'
+>;
+
+export interface ReactionsProps extends Pick<PaletteProps, 'size' | 'disabled'>, QAProps, DOMProps {
```
Why we need `disabled`? What is use case for that?

amje replied:
Now disabled buttons look barely visible, I think it's a bad UX

Ruminat replied:
Do we have a button state for that?
Because simply disabling the `onClickReaction` callback looks like a bad UX as well: a user hovers a button, it reacts (the background changes), he clicks on it, the button reacts again (it changes it's size slightly and goes back) and then simply nothing happens — that would be quite confusing

Ruminat commented:
> @Ruminat Let's simplify the API, currently it looks a bit hard to understand and it stores the state in multiple places (`reactions` and `palette.options`). What do you think about this API suggestion?
> 
> ```ts
> // Do not extend whole PaletteOption, only pick needed props. This makes refactor easier
> interface ReactionsItem extends Pick<PaletteOption, 'value' | 'content' | 'title'> {
>     count: number;
>     selected?: boolean;
> }
> 
> interface ReactionsProps {
>     // This would map into Palette props (options and value depending on selected states) and create buttons if count > 0
>     reactions: ReactionsItem[];
>     // User should handle this callback and change reactions array
>     onToggle?: (reaction: ReactionsItem) => void;
>     // Extract tooltip render to dedicated prop and pass whole reaction item
>     // No other tooltip props needed for the sake of simplicty
>     renderTooltipContent?: (reaction: ReactionsItem) => React.ReactNode;
>     // Keep only columns. Other Palette props are useless
>     columns?: number;
>     // Rest props
>     // ...
> }
> ```
> 
> The only source of state is now the `reactions` array, which user have to update on `onToggle` calls. I also extracted tooltip prop of reaction into top-level prop `renderTooltipContent`. Copying the same tooltip props for each reaction is redundant.

Yeah, I like that solution, I'll change the code accordingly

Ruminat commented:
@amje  I started implementing your solution and encountered a problem:

let's say you have three reactions: 😊, 👍 and 😢
and at first you have this state: 😊:0 👍:0 😢:0 and the user sees no reactions
the user clicks on 😢, so the state becomes 😊:0 👍:0 😢:1, the users sees 😢:1
and when the users clicks on 😊, instead of seeing 😢:1 😊:1 he sees 😊:1 😢:1., because we just increase a counter and don't change the emoji's position
you can change the position though, so the user sees 😢:1 😊:1, but in such a case `Palette`'s order of options will change accordingly

So I guess we have to use 2 states here

Ruminat commented:
I've also noticed this comment:
```typescript
// Extract tooltip render to dedicated prop and pass whole reaction item
// No other tooltip props needed for the sake of simplicty
```
Tooltip is a little bit harder than just rendering content, tooltip has the following type:
```typescript
export interface ReactionTooltipProps
    extends Pick<PopoverProps, 'strategy' | 'placement' | 'modifiers'> {
    /**
     * Tooltip's content.
     */
    content: React.ReactNode;
    /**
     * Tooltip content's HTML class attribute.
     */
    className?: string;
    /**
     * Fires when the `onMouseLeave` callback is called.
     * Usage example:
     * you have some popup inside a tooltip, you hover on it, you don't want the tooltip to be closed because of that.
     */
    canClosePopup?: () => boolean;
}
```

`className` might not be important, but `canClosePopup` can be. In Arcanum, for example, it is used the block the tooltip's closure on hovering on a user card popup. `Pick<PopoverProps, 'strategy' | 'placement' | 'modifiers'>` can be important as well because `uikit`'s Popups can act strange sometimes. For example, in Nirvana we had to hack it a bit so it doesn't go over the visible part of the screen.

Ruminat commented on src/components/Reactions/mock/mockData.ts:
fixed

Ruminat commented:
@amje ping

Ruminat commented:
@amje ping

amje commented:
> I've also noticed this comment:
> 
> ```ts
> // Extract tooltip render to dedicated prop and pass whole reaction item
> // No other tooltip props needed for the sake of simplicty
> ```
> 
> Tooltip is a little bit harder than just rendering content, tooltip has the following type:
> 
> ```ts
> export interface ReactionTooltipProps
>     extends Pick<PopoverProps, 'strategy' | 'placement' | 'modifiers'> {
>     /**
>      * Tooltip's content.
>      */
>     content: React.ReactNode;
>     /**
>      * Tooltip content's HTML class attribute.
>      */
>     className?: string;
>     /**
>      * Fires when the `onMouseLeave` callback is called.
>      * Usage example:
>      * you have some popup inside a tooltip, you hover on it, you don't want the tooltip to be closed because of that.
>      */
>     canClosePopup?: () => boolean;
> }
> ```
> 
> `className` might not be important, but `canClosePopup` can be. In Arcanum, for example, it is used to block the tooltip's closure on hovering on a user card popup. `Pick<PopoverProps, 'strategy' | 'placement' | 'modifiers'>` can be important as well because `uikit`'s Popups can act strange sometimes. For example, in Nirvana we had to hack it a bit so it doesn't go over the visible part of the screen.

You are trying to fix those issues implementing extra stuff in the component that not needed. It's better to have simple and robust API that would work predictable.

Ruminat commented:
@amje I simplified `ReactionProps`'s `tooltip` to simply `React.ReactNode` and removed `content` and `title` props from `ReactionProps` (moved them to `ReactionInnerProps`, so they are calculated in the `Reactions` component)

Ruminat commented:
@amje ping

amje commented:
@Ruminat As we discussed eariler, let's implement the following API:

```ts
interface Reaction {
  value: string; // Unique identifier of reaction
  content: React.ReactNode; // Reaction's content
  title?: string; // Alternative content via browser's tooltip
}

interface ReactionState {
  value: string;
  selected?: boolean;
  count?: number;
}

interface ReactionsProps extends QAProps, DOMProps {
  reactions: Reaction[]; // <-- Keep all data in single array
  reactionsState?: ReactionState[]; // <-- Keep all state in place
  onToggle?: (value: string) => void; // User should update reactionsState in this callback
  // Other props
  readOnly?: boolean; // Instead of disabling buttons and coloring them to gray we hide palette trigger and ignore hover/click events
  paletteProps?: Pick<PaletteProps, 'columns' | 'rowClassName' | 'optionClassName'> // Only pick props related to view
}
```

It's robust, simple and easy to understand.

amje commented on src/components/Reactions/hooks.ts:
```diff
@@ -0,0 +1,88 @@
+        setCurrentHoveredReaction({reaction, open: true, ref});
+    });
+
+    const close = useStableCallback(() => {
```
As Alexey mentioned above we can just use useCallback instead. Also the implementation of `useStableCallback` violates React's rules (writing or reading refs during render)

Ruminat replied:
Removed `useStableCallback` completely

Ruminat commented:
@amje I implemented your suggested API

Ruminat commented:
@amje I also added `tooltipBehavior` property to `Reactions` component:
It defined how a reaction's tooltip should act:
1. as a tooltip (you can't hover over the contents — it disappeares),
2. as a popover (you can hover over the contents — it doesn't disappear).

What do you think?

Ruminat commented:
I also added myself as owner of `Notifications` and `Reactions` components

Ruminat commented:
@amje ping

Ruminat commented:
@amje ping

amje commented on src/components/Reactions/Reaction.tsx:
```diff
@@ -0,0 +1,105 @@
+interface ReactionInnerProps extends Pick<PaletteOption, 'content'> {
+    reaction: ReactionStateProps;
+    size: ButtonSize;
+    tooltipBehavior: 'tooltip' | 'popover';
```
Do we really need that prop? Why it can't be default as popover?

Ruminat replied:
I'm not sure
In our services we always use the `popover` behavior
But someone might not need it. For example: someone can simply show users' avatars without any interactivity

Ruminat replied:
Removed the prop

amje commented on src/components/Reactions/Reaction.tsx:
```diff
@@ -0,0 +1,105 @@
+
+    const button = (
+        <span
+            onMouseEnter={props.tooltipBehavior === 'tooltip' ? onMouseEnter : undefined}
```
We can assing handlers directly to the button, no need to add wrapper

Ruminat replied:
fixed

Ruminat commented on src/components/Reactions/Reactions.scss:
```diff
@@ -0,0 +1,49 @@
+
+#{$block} {
+    &__reaction-button_add-button {
+        color: var(--g-color-text-secondary);
```
I remember having problems using `Text` inside a `Button`

<img width="303" alt="image" src="https://github.com/user-attachments/assets/a96ca8d1-d870-4e69-b851-7371969ea285">

So I'm not sure if that's a good idea

Ruminat replied:
Doesn't look like it:

<img width="640" alt="image" src="https://github.com/user-attachments/assets/2293964e-9503-4c10-addf-1076691baff1">

Icon with text seems to be a little higher than it should be
The problem (If I remember corretly) lies in the `Text`'s `line-height`

Ruminat replied:
A styled `<span />` worked just fine though
So maybe let's just stick to that?

Ruminat replied:
Used `flat-secondary` view instead

Ruminat commented on src/components/Reactions/Reaction.tsx:
```diff
@@ -0,0 +1,105 @@
+     * If present, when a user hovers over the reaction, a popover appears with `tooltip.content`.
+     * Can be used to display users who used this reaction.
+     */
+    tooltip?: React.ReactNode;
```
The `reactions` data array is all of the available reactions
And this `tooltip` is used to show reacted users, so it's shown in the used reactions
So I don't see how we can move to the `reactions` array

amje replied:
Oh, I see. Then maybe move it to the separete top-level prop `renderTooltip(state: ReactionState)`?

Ruminat replied:
Take a look at usage with `renderTooltip`:

<img width="1276" alt="image" src="https://github.com/user-attachments/assets/40907932-101f-4468-84f4-d7699a80567f">

Isn't it more complicated?

Ruminat replied:
Turned `tooltip` into `renderTooltip`

Ruminat commented on src/components/Reactions/Reactions.tsx:
```diff
@@ -0,0 +1,174 @@
+                        <Button
+                            className={b('reaction-button', {size, 'add-button': true})}
+                            size={size}
+                            extraProps={{'aria-label': 'add-reaction'}}
```
fixed

Ruminat commented on src/components/Reactions/Reactions.tsx:
```diff
@@ -0,0 +1,174 @@
+                        content={paletteContent}
+                        tooltipContentClassName={b('add-reaction-popover')}
+                        openOnHover={false}
+                        hasArrow={false}
```
fixed

Ruminat commented on src/components/Reactions/Reactions.scss:
```diff
@@ -0,0 +1,49 @@
+    &__reaction-button_size_l {
+        font-size: 16px;
+    }
+    &__reaction-button_size_xl {
```
fixed

Ruminat commented on src/components/Reactions/Reaction.tsx:
```diff
@@ -0,0 +1,105 @@
+
+export type ReactionProps = Pick<PaletteOption, 'value' | 'content' | 'title'>;
+
+export interface ReactionStateProps {
```
fixed

Ruminat commented:
@amje ping

Ruminat commented:
@amje ping

Ruminat commented:
@amje ping
